src/Orleans/SystemTargetInterfaces/IRemoteGrainDirectory.cs

```diff
@@ -22,7 +22,7 @@ internal interface IGrainInfo
         bool AddActivation(ActivationId act, SiloAddress silo);
         ActivationAddress AddSingleActivation(GrainId grain, ActivationId act, SiloAddress silo, GrainDirectoryEntryStatus registrationStatus);
         bool RemoveActivation(ActivationId act, UnregistrationCause cause, out IActivationInfo entry, out bool wasRemoved);
-        bool Merge(GrainId grain, IGrainInfo other);
+        bool Merge(GrainId grain, IGrainInfo other, IInternalGrainFactory grainFactory);
```

Comment:
Because this object is allocated for each Grain which has an activation, I opted for this approach. Alternatives would be:
* Make it a class member
* Move the method to another type
* Avoid the need for it

I think this is cleaner than the first two. The latter could be implemented by changing the contract of `IGrainInfo.Merge` to return a map of `silos -> activations to drop` and having the caller deactivate them.